enhancements/api-review/add-new-CRD-ImageContentPolicy(ICP)-to-config.openshift.io.md

```diff
@@ -0,0 +1,325 @@
+
+### Goals
+
+- Enable the user to pull images using tags from mirrors configured through config.openshift.io/v1 API ImageContentPolicy.
+
+### Non-Goals
+
+- This proposal does not recommend using by-tag references for OpenShift release images. Those should still be referenced by digest, regardless of whether `MirrorsByTags` is configured for repositories where the release images are mirrored
+
+## Proposal
+
+The New CRD ImageContentPolicy should be added to config.openshift.io/v1.
+
+The schema of ImageContentPolicy(ICP) = fields of ImageContentSourcePolicy(ICSP) + MirrorsByTags
+
+The `MirrorsByTags` will allow for pulling images by tag through configured mirrors.
+
+```go
+// ImageContentPolicySpec is the specification of the ImageContentPolicy CRD.
+type ImageContentPolicySpec struct {
+	// repositoryMirrors allows images referenced by image digests in pods to be
+	// pulled from alternative mirrored repository locations. The image pull specification
+	// provided to the pod will be compared to the source locations described in RepositoryMirrors
+	// and the image may be pulled down from any of the mirrors in the list instead of the
+	// specified repository allowing administrators to choose a potentially faster mirror.
+	// To pull image from mirrors using tag specification, should configure 
+	// a list of mirrors to "mirrorsByTags".
+	//
+	// If the image pull specification matches the reposirtory of "source" in multiple policies,
+	// only the policies have the most specific namespace matching will be used.
+	// For example, if there are policies using quay.io/libpod and quay.io/libpod/busybox as
+	// the "source", only the policies using quay.io/libpod/busybox are going to apply
+	// for pull specification quay.io/libpod/busybox.
+	// Each “source” repository is treated independently; configurations for different “source”
+	// repositories don’t interact.
+	//
+	// If the "mirrors" or "mirrorsByTags" is not specified, the image will continue to be pulled from the specified
+	// repository in the pull spec.
+	//
+	// When multiple policies are defined for the same “source” repository, the sets of defined
+	// mirrors(mirrors configured in the same field, "mirrors" or "mirrorsByTgas") will be merged together, preserving the relative order of the mirrors, if possible.
+	// For example, if policy A has mirrors `a, b, c` and policy B has mirrors `c, d, e`, the
+	// mirrors will be used in the order `a, b, c, d, e`.  If the orders of mirror entries conflict
+	// (e.g. `a, b` vs. `b, a`) the configuration is not rejected but the resulting order is unspecified.
+	// If want to use a deterministic order of mirrors, configure them into one list of mirrors using expected order.
+	// +optional
+	// +listType=atomic
+	RepositoryMirrors []RepositoryMirrors `json:"repositoryMirrors"`
+}
+
+// RepositoryMirrors holds cluster-wide information about how to handle mirrors in the registries config.
```

Comment:
Can we make the CRD namespaced? 

I don't want to hook too much into this proposal since it stands for its own. Still, there is another feature request which aims to make the mirror configuration possible from an OpenShift project (namespace): https://issues.redhat.com/browse/OCPNODE-765

Having the CRD in a namespaced scope would help us to avoid duplication and simplify the integration into the MCO.

Reply:
Yes, I am thinking about it too. Would this impact the process to update the components that use ICSP now to use new CRD? They will use the default namespace after using this new CRD?

Reply:
- AFAICS https://issues.redhat.com/browse/OCPNODE-765 does _not_ ask for namespacing the mirror configuration.
- If we did support per-namespace mirror configuration (assuming that were in addition to the existing cluster-wide functionality, which probably needs to continue to exist without per-namespace opt-in), wouldn’t it be a namespaced CR, conceptually a different CRD from this cluster-wide resource?

Reply:
I think @mtrmac is right and I misinterpreted the request. This would mean that it has no use any more to make it a namespaced resource.

Reply:
Sure. I'd remove the change to namespaced in this enhancement.